### ISSUE TITLE
ossec-authd segfaults on agent registration

### ISSUE THREAD

**Issue by copacetic** [wazuh/wazuh#148]
This occurs on `64-bit Debian Wheezy`. I used the install instructions here: https://documentation.wazuh.com/2.0/installation-guide/installing-wazuh-agent/wazuh_agent_deb.html

Output from the Server's command `sudo strace /var/ossec/bin/ossec-authd -d -x /var/ossec/etc/sslmanager.cert -k /var/ossec/etc/sslmanager.key`:

```
...
2017/06/29 16:57:46 ossec-authd: INFO: Received request for a new agent (example.com) from: *.*.*.*
2017/06/29 16:57:46 ossec-authd: INFO: Agent key generated for example.com (requested by *.*.*.*)
2017/06/29 16:57:46 ossec-authd: INFO: Agent key created for example.com (requested by *.*.*.*)
 <unfinished ...>
+++ killed by SIGSEGV +++
```

Both client and server appear to show everything is copacetic. You only see the SIGSEGV if you use strace.
Client output:

```hayg@Client:~$ sudo /var/ossec/bin/agent-auth -d -A Client -m Server -v /var/ossec/etc/rootCA.pem
2017/06/29 16:57:46 agent-auth: DEBUG: Starting ...
2017/06/29 16:57:46 agent-auth: INFO: Started (pid: 50547).
2017/06/29 16:57:46 ossec-authd: DEBUG: Peer verification requested.
2017/06/29 16:57:46 ossec-authd: DEBUG: Returning CTX for client.
WARN: No authentication password provided.
INFO: Connected to *.*.*.*:1515
INFO: Verifing manager's certificate
2017/06/29 16:57:46 ossec-authd: DEBUG: Checking certificate's subject alternative names.
INFO: Using agent name as: example.com
INFO: Send request to manager. Waiting for reply.
INFO: Received response with agent key
INFO: Valid key created. Finished.
INFO: Connection closed.
```
Looking at it in `gdb`, I get the following `bt`:
```Program received signal SIGSEGV, Segmentation fault.
[Switching to Thread 0xf73bab70 (LWP 24534)]
strlen () at ../sysdeps/i386/i486/strlen.S:69
69      ../sysdeps/i386/i486/strlen.S: No such file or directory.
(gdb) bt
#0  strlen () at ../sysdeps/i386/i486/strlen.S:69
#1  0xf7c83935 in *__GI___strdup (s=0x0) at strdup.c:42
#2  0x08050c37 in ?? ()
#3  0x0804df72 in ?? ()
#4  0xf7d63954 in start_thread (arg=0xf73bab70) at pthread_create.c:304
#5  0xf7ce4d1e in clone () at ../sysdeps/unix/sysv/linux/i386/clone.S:130
```

This is a huge blocker for me and I'm certain for others as well. 

EDIT: Adding dpkg output
```
ii  wazuh-manager                               2.0-1wheezy                         i386         Wazuh helps you to gain security visibility into your infrastructure by monitoring hosts at an operating system and application level. It provides the following capabilities: log analysis, file integrity monitoring, intrusions detection and policy and compliance monitoring
```

**Comment by vikman90** [comment]
Hi @copacetic,

thanks for notifying us. We are trying to reproduce this problem but we still have not got it.

First, could you confirm what architecture you have using? You mentioned `64-bit` Debian but both Wazuh package and gdb output refer to `i386`.

Please help us with these questions:

- How many agents are registered when `ossec-authd` starts? (`wc -l /var/ossec/etc/client.keys`)
- Can you confirm that although `strace` notifies that error, the key was properly installed? In this case, the last line of file `/var/ossec/etc/client.keys` of the manager would match the only line in the same file of the agent.

I suspect that the `strdup()` problem may be here: https://github.com/wazuh/wazuh/blob/stable/src/os_auth/main-server.c#L779 but I did not find any issue in that part of the code.

Thank you again.
Best regards.

**Comment by copacetic** [comment]
Thank you for the prompt reply, vikman90.

I said 64 bit based on our `uname` output:
```
hayg@Server:~/ossec_tmp/ossec-wazuh$ uname -r
3.2.0-4-amd64
```
This machine is actually a 64 bit kernel, but runs multiarch userspace in both 32 bit and 64 bit. I wonder if maybe that's causing this somehow.

**Comment by copacetic** [comment]
Small update. I compiled the package on the Server machine and it worked. I think this means ossec-authd can't handle the mix of 64 bit kernel / 32 bit userspace gracefully. I'm sure it can probably be easily fixed to work in this environment or at least detect it.

**Comment by copacetic** [comment]
Although the server did not crash, it's failing to talk to the server.
```
2017/06/30 16:27:05 agent-auth: DEBUG: Starting ...
2017/06/30 16:27:05 agent-auth: INFO: Started (pid: 35657).
2017/06/30 16:27:05 ossec-authd: DEBUG: Peer verification requested.
2017/06/30 16:27:05 ossec-authd: DEBUG: Returning CTX for client.
2017/06/30 16:27:05 ossec-authd: DEBUG: Checking certificate's subject alternative names.
2017/06/30 16:32:23 ossec-agentd: INFO: Trying to connect to server (10.128.77.130:1514).
2017/06/30 16:32:44 ossec-agentd(4101): WARN: Waiting for server reply (not started). Tried: '10.128.77.130'.
2017/06/30 16:42:45 ossec-agentd: INFO: Trying to connect to server (10.128.77.130:1514).
2017/06/30 16:43:06 ossec-agentd(4101): WARN: Waiting for server reply (not started). Tried: '10.128.77.130'.
2017/06/30 16:53:07 ossec-agentd: INFO: Trying to connect to server (10.128.77.130:1514).
2017/06/30 16:53:28 ossec-agentd(4101): WARN: Waiting for server reply (not started). Tried: '10.128.77.130'.
```
`netstat` on the server shows no connection is established, but that it's listening on port 1514. Firewall is not an issue, I confirmed via `tcpdump` on the Server that it is receiving the UDP traffic on port 1514. The `client.keys` file has a matching secret on both the Server and Client.

**Comment by copacetic** [comment]
I built the agent on the Client machine and tried registering via `ossec_authd`. I hit a segfault this time. Very confused.

**Comment by vikman90** [comment]
Hi,

We are still trying to reproduce the error. I configured multiarchitecture in `dpkg` and installed package `wazuh-manager:i386` in Debian Wheezy _amd64_.

The `ossec-authd` server appears to work correctly and neither `strace` nor `gdb` nor `valgrind` report any problem.

Please try to uninstall every Wazuh package from your machine and remove `/var/ossec`. Then install explicitly `wazuh-manager:amd64`.

Regarding to building from sources, it would be very interesting that you do it with debugging enabled. In order to achieve this:

```
# cd wazuh-source-folder
sudo make -C src clean
make -C src TARGET=server DEBUG=1
sudo ./install.sh
```

If `wazuh-authd` crashes again, try to launch it with Valgrind:

```
valgrind --leak-check=full --trace-children=yes /var/ossec/bin/ossec-authd -d
```

This way Valgrind should give the code line that was being executed in the moment of crashing.

I hope we can solve this problem.
Best regards.

**Comment by copacetic** [comment]
Sorry I've been on holiday. Thanks for the advice. Here's my stack trace with debugging enabled:
```
(gdb) bt
#0  strlen () at ../sysdeps/i386/i486/strlen.S:69
#1  0xf7c83935 in *__GI___strdup (s=0x0) at strdup.c:42
#2  0x08050c37 in OS_DupKeys ()
#3  0x0804df72 in run_writer ()
#4  0xf7d63954 in start_thread (arg=0xf73bab70) at pthread_create.c:304
#5  0xf7ce4aae in clone () at ../sysdeps/unix/sysv/linux/i386/clone.S:130
```

**Comment by copacetic** [comment]
Relevant valgrind output:
```
... Agent key created ...                             
==31892== Warning: set address range perms: large range [0x38e55028, 0x590346ac) (defined)
==31892== Thread 3:
==31892== Use of uninitialised value of size 4
==31892==    at 0x5009A48: strlen (mc_replace_strmem.c:390)
==31892==    by 0x71C4934: strdup (strdup.c:42)
==31892==    by 0x8050C36: OS_DupKeys (in /var/ossec/bin/ossec-authd)
==31892==    by 0x804DF71: run_writer (in /var/ossec/bin/ossec-authd)
==31892==    by 0x7225AAD: clone (clone.S:130)
==31892==
==31892== Invalid read of size 1
==31892==    at 0x5009A48: strlen (mc_replace_strmem.c:390)
==31892==    by 0x71C4934: strdup (strdup.c:42)
==31892==    by 0x8050C36: OS_DupKeys (in /var/ossec/bin/ossec-authd)
==31892==    by 0x804DF71: run_writer (in /var/ossec/bin/ossec-authd)
==31892==    by 0x7225AAD: clone (clone.S:130)
==31892==  Address 0x0 is not stack'd, malloc'd or (recently) free'd
==31892==
==31892==
==31892== Process terminating with default action of signal 11 (SIGSEGV)
==31892==  Access not within mapped region at address 0x0
==31892==    at 0x5009A48: strlen (mc_replace_strmem.c:390)
==31892==    by 0x71C4934: strdup (strdup.c:42)
==31892==    by 0x8050C36: OS_DupKeys (in /var/ossec/bin/ossec-authd)
==31892==    by 0x804DF71: run_writer (in /var/ossec/bin/ossec-authd)
==31892==    by 0x7225AAD: clone (clone.S:130)
```

It's worth noting this is in EC2. Also, it seems to have trouble finding the client.keys:
`2017/07/05 12:46:00 ossec-authd(1751): ERROR: File client.keys not found or empty.`

**Comment by copacetic** [comment]
Installing the `amd64` package isn't an option because of dependency hell. It would remove 32-bit versions of core packages, like cron.

**Comment by copacetic** [comment]
Could it be that this is all being caused by `client.keys` being empty/not found?
I think the trace starts here, in `src/os_auth/main-server.c`:
```
...
735        copy_keys = OS_DupKeys(&keys);
...
```
I'm imagining keys is `null` and `OS_DupKeys` segfaults if that's the case. The questions then are:
1. Why is it `null`?
2. Should we be segfaulting when it's `null`?

**Comment by vikman90** [comment]
Hi,

I'm reviewing the code for the stable branch (version 2.0.0). The `strdup()` may be one of these: https://github.com/wazuh/wazuh/blob/stable/src/os_crypto/shared/keys.c#L584.

All of them are protected with an `if` statement that guards that the pointer is not `null`. This should happen only for the last item (that acts as an array end mark).

Then the program checks `ip` but duplicates `ip->ip`. The IP is parsed by https://github.com/wazuh/wazuh/blob/stable/src/shared/validate_op.c#L293, that guarantees that `ip->ip` won't be null if it returns non-zero.

I tested it once again and got no problems. Don't worry about the `File client.keys not found or empty` error, it only appears when using option `-d`, it should not be related to this problem.

Since the problem appears only in your environment, I would be very pleased if you followed these points:

1. Did you compile with `DEBUG=1`? This option should make the exact line of code to appear in `gdb` and `valgrind`. I'd like to know which of these `strdup()` is leading to the segmentation fault.

2. Please prevent removed keys from being copied by running Authd with option `-r`.

3. Fill the file `/var/ossec/etc/client.keys` with whatever key, for instance adding this line:

> 001 myagent any abcdefghijklmnopqrstuvwxyz

Thank you again for your feedback.
Best regards.

**Comment by copacetic** [comment]
I did recompile with `DEBUG=1` but I didn't delete the old installation. It asked me if I want to "update" and I said yes. If there's an easy way to uninstall, I can reinstall from scratch. Maybe updating didn't overwrite the binaries?

I used `-r` and changed the contents of `client.keys` to be what you pasted. It actually worked in that the `authd` server didn't crash! However, I still haven't been able to establish a connection. Note that I was able to establish a connection when I manually copy pasted the secret key from server to agent.

The logs look like so:
```
2017/07/05 14:38:46 ossec-agentd(4101): WARN: Waiting for server reply (not started). Tried: 'Server'.
2017/07/05 14:39:24 ossec-agentd: INFO: Trying to connect to server (Server:1514).
```

**Comment by vikman90** [comment]
Glad to know!

Authd registers an agent but does not save the server's address. You can configure it by editing the file `/var/ossec/etc/ossec.conf` at `<server-ip>`. Fill it with your manager's IP (probably the same address that you used with `agent-auth`) and restart the agent.

This documentation may help you: [Client documentation](https://documentation.wazuh.com/current/user-manual/reference/ossec-conf/client.html).

Please repeat the same steps with and without option `-r`. If the program crashes only when installed from packages (they were compiled from branch _stable_) maybe the problem is in the package build.

Best regards.

**Comment by vikman90** [comment]
If the agent is able to connect when copying the secret, retry to register with Authd and check that both files `/var/ossec/etc/client.keys` were updated. Please let me know if one of them was not updated.

**Comment by copacetic** [comment]
Aha, it seems that setting `<server-hostname>` instead of `<server-ip>` fixed it. Thanks for that! I'll toggle `-r`.

**Comment by copacetic** [comment]
It segfaults without the `-r`. I can try running the server from the package, but I got the server running from source to segfault without the `-r` so it seems unnecessary. What do you think?

**Comment by vikman90** [comment]
Hi,

I think I've fixed it! I managed to reproduce the issue and the pointer to the "deleted keys" was incorrect.

The problem was in the interpretation of `ino_t` at https://github.com/wazuh/wazuh/blob/2.0/src/headers/sec.h#L51: `main-server.c` interpreted it as a 32-bit integer but `keys.c` interpreted it as a 64-bit integer.

I found that there is a strange definition at https://github.com/wazuh/wazuh/blob/2.0/src/headers/shared.h#L34 to use 64-bit values for files due to the ZLib library configuration. But `main-server.c` included `sys/types.h` before defining that value and this made `ino_t` be a 32-bit type.

In other words, the issue was due to the order in which header files are included. I have fixed it (https://github.com/wazuh/wazuh/commit/7e5446dba89e0e3ff958b576da052d35b7122e23) and created a branch [debug-authd](https://github.com/wazuh/wazuh/tree/debug-authd) with more _debug2_ messages (they will be shown with option `-d -d`) and it includes this fix.

I would be very pleased if you could download that branch, or [branch 2.0](https://github.com/wazuh/wazuh/tree/debug-authd), and confirm me that the problem is solved.

Thank you very much, your help has been very useful to solve this problem.
Best regards.

**Comment by copacetic** [comment]
That's awesome. Great work :) I'll try and test this when I have some free cycles. In either case, you fixed a real segfault, for sure.

**Comment by vikman90** [comment]
Thank you again @copacetic,

let me close this issue. If you find any problem in your tests please reopen it.

Regards.

**Comment by copacetic** [comment]
@vikman90  I just wanted to let you know that your hard and high quality work has paid off. I can confirm the segfault is no longer an issue for me.

**Comment by vikman90** [comment]
Hi @copacetic,

I'm glad to know that this problem was solved! Thank you again for your feedback, we appreciate your help and your comments.
